Sync v2-9-stable with v2-9-test to release python client v2.9.0

ephraimbuddy opened the pull request:
Time for python client `2.9.0rc1`!

potiuk committed:
Hide internals of some groupped workflows (#38604)

When the whole workflow can be skipped and it's one of the workflows
than nothing depends on, we can skip the whole workflow rather than
individual jobs - then the workflow will not be shown as unfolded
with all the jobs skipped but as a single skipped workflow.

This is what we do for Additional CI image tests and Finalize tests
& special tests.

(cherry picked from commit 3f4872beea156a60e7318b806c36be53a05207a9)

jedcunningham committed:
Fix kubeconform pre-commit to not leave around containers (#38688)

(cherry picked from commit d212b119e1bf9930a4fb444b1667e388f093e236)

potiuk committed:
Workaround occasional failures of docker system prune in public runners (#38825)

Occasionally `docker syetem prune` fails on Public docker runners with

file resize error: truncate /var/lib/docker/buildkit/containerdmeta.db:
read-only file system

This happens after docker has been already cleaned and should have no
side efects if the metadata db could not be truncated.

Ignoring error in this case should allow the job to continue.

(cherry picked from commit f50cc6180c28d2cbfd47131ec4808363d9a857e2)

ephraimbuddy committed:
Add FAB provider to Airflow extras in the image (#38824)

(cherry picked from commit e77c1358eb3d9ce19806ed8918ce5825878eca66)

ephraimbuddy committed:
Update Airflow version in Dockerfile

potiuk committed:
Update build dependencies to their latest versions (#38930)

(cherry picked from commit 3af00fa2a206e31a449648e33eed1f806d8b62b7)

Taragolis committed:
Mark `extending/add-build-essential-extend/Dockerfile` docker example test as XFAIL (#38978)

(cherry picked from commit 70e17f6e0010e369a4b2d05869c66ff902c3ba42)

potiuk committed:
Fix PROD image package installation in CI (#39035)

When PROD image packages are installed in in CI, the local sources
should not be present in the image, also constraints from sources
shoudl replace the one downloaded from main.

(cherry picked from commit 246f69787358dd63df8d91dbba14ee29aecddb84)

potiuk committed:
Do not add local airflow sources in CI when building k8s image (#39043)

Follow up after #39035 where local airflow sources have been
removed from PROD image built in CI. The k8s image for tests also
added the sources when preparing the modified image. This is now
avoided - the --no-copy-local-sources is used in CI to avoid it.

(cherry picked from commit 1e9d3d1c351ce10e0b41db6e22b4fb9ad6537703)

hussein-awala committed:
update simple-salesforce type hints to support 1.12.6 (#39047)

(cherry picked from commit 30250962db2404559190b9f2e5bb55652deaae21)

ephraimbuddy committed:
Add changelog for airflow python client 2.9.0 (#39060)

(cherry picked from commit 4d94df489a92379428ca5ab684a4bf75b6522b9c)

potiuk committed:
Upgrade to latest hatchling as build dependency (#39044)

(cherry picked from commit 6995eafc8ae2b9bfd1e75712bba6756e9c69f68f)

potiuk reviewed (approved):
It will need one or two cherry-picks more from main to fix but LGTM in general.

ephraimbuddy committed:
Update hatchling to version 1.24.0 (#39072)

(cherry picked from commit 1d34045f3aa41bad8cc05bdc1391ab1edcdf3ad5)

potiuk committed:
Fix main failing because of the new xmlsec 1.3.14 compatibility (#39104)

The xmlsec used by Amazon provider for authentication has compatiblity
issue with libxmlsec 1.2.* that is used by default in debian bookworm.

We should investigate if/howe we can upgrade our images to support it,
in the meantime we limitt python bindings to < 1.3.13.

See https://github.com/apache/airflow/issues/39103

(cherry picked from commit cd888870ba1b95c20208b1c87d337cd2418e1dc1)

potiuk committed:
Upgrade to latest hatchling 1.24.1 (again).

(cherry picked from commit 43418928b86a7de891f5b9c1abf50cf0c3a9941b)

potiuk committed:
Apply PROVIDE_PROJECT_ID mypy workaround across Google provider (#39129)

There is a simple workaround implemented several years ago for Google
provider `project_id` default value being PROVIDE_PROJECT_ID that
satisfy mypy checks for project_id being set. They way how
`fallback_to_default_project_id` works is that across all the
providers the project_id is actually set, even if technically
it's default value is set to None.

This is similar typing workaround as we use for NEW_SESSION in the
core of Airflow.

The workaround has not been applied consistently across all the
google provider code and occasionally it causes MyPy complaining
when newer version of a google library introduces more strict
type checking and expects the provider_id to be set.

This PR applies the workaround across all the Google provider
code.

This is - generally speaking a no-op operation. Nothing changes,
except MyPy being aware that the project_id is actually going to
be set even if it is technically set to None.

(cherry picked from commit 90acbfbba1a3e6535b87376aeaf089805b7d3303)
